src/libmongoc/src/mongoc/mongoc-server-monitor.c

```diff
@@ -0,0 +1,1250 @@
+#include "mongoc/mongoc-topology-background-monitoring-private.h"
+#include "mongoc/mongoc-topology-private.h"
+#include "mongoc/mongoc-trace-private.h"
+
+#undef MONGOC_LOG_DOMAIN
+#define MONGOC_LOG_DOMAIN "monitor"
+
+typedef enum {
+   MONGOC_THREAD_OFF = 0,
+   MONGOC_THREAD_RUNNING,
+   MONGOC_THREAD_SHUTTING_DOWN,
+   MONGOC_THREAD_JOINABLE
+} thread_state_t;
+
+/* Use a signed and wide return type for timeouts as long as you can. Cast only
+ * when you know what you're doing with it. */
+static int64_t
+_now_us (void)
+{
+   return bson_get_monotonic_time ();
+}
+
+static int64_t
+_now_ms (void)
+{
+   return _now_us () / 1000;
+}
+
+struct _mongoc_server_monitor_t {
+   mongoc_topology_t *topology;
+   bson_thread_t thread;
+
+   /* State accessed from multiple threads. */
+   struct {
+      bson_mutex_t mutex;
+      mongoc_cond_t cond;
+      thread_state_t state;
+      bool scan_requested;
+      bool cancel_requested;
+   } shared;
+
+   /* Default time to sleep between ismaster checks (reduced when a scan is
+    * requested) */
+   uint64_t heartbeat_frequency_ms;
+   /* The minimum time to sleep between ismaster checks. */
+   uint64_t min_heartbeat_frequency_ms;
+   int64_t connect_timeout_ms;
+   bool use_tls;
+#ifdef MONGOC_ENABLE_SSL
+   mongoc_ssl_opt_t *ssl_opts;
+#endif
+   mongoc_uri_t *uri;
+   /* A custom initiator may be set if a user provides overrides to create a
+    * stream. */
+   mongoc_stream_initiator_t initiator;
+   void *initiator_context;
+   int64_t request_id;
+   mongoc_apm_callbacks_t apm_callbacks;
+   void *apm_context;
+
+   mongoc_stream_t *stream;
+   bool more_to_come;
+   mongoc_server_description_t *description;
+   uint32_t server_id;
+   bool is_rtt;
+};
+
+static void
+_server_monitor_log (mongoc_server_monitor_t *server_monitor,
+                     mongoc_log_level_t level,
+                     const char *format,
+                     ...) BSON_GNUC_PRINTF (3, 4);
+
+static void
+_server_monitor_log (mongoc_server_monitor_t *server_monitor,
+                     mongoc_log_level_t level,
+                     const char *format,
+                     ...)
+{
+   va_list ap;
+   char *msg;
+
+   va_start (ap, format);
+   msg = bson_strdupv_printf (format, ap);
+   va_end (ap);
+
+   mongoc_log (level,
+               MONGOC_LOG_DOMAIN,
+               "[%s%s] %s",
+               server_monitor->description->host.host_and_port,
+               server_monitor->is_rtt ? "-RTT" : "",
+               msg);
+   bson_free (msg);
+}
+
+#ifdef MONGOC_TRACE
+#define MONITOR_LOG(sm, ...) \
+   _server_monitor_log (sm, MONGOC_LOG_LEVEL_TRACE, __VA_ARGS__)
+#else
+#define MONITOR_LOG(sm, ...) (void) 0
+#endif
+
+/* TODO use MONGOC_LOG_LEVEL_ERROR */
+#define MONITOR_LOG_ERROR(sm, ...) \
+   _server_monitor_log (sm, MONGOC_LOG_LEVEL_DEBUG, __VA_ARGS__)
+/* TODO use MONGOC_LOG_LEVEL_WARNING */
+#define MONITOR_LOG_WARNING(sm, ...) \
+   _server_monitor_log (sm, MONGOC_LOG_LEVEL_DEBUG, __VA_ARGS__)
```

Comment:
There are assertions in evergreen that nothing to stderr is printed in tests without being captured within the test. Since there are quite a few tests that use the mock server to simulate network errors on monitoring, I'd prefer doing this in a follow-up PR if that is ok with ya'll.